BREAKING CHANGE: refactor PackageURL by moving String functions to StringUtil

jeremylong opened the pull request:
- Move string validation and encoding functions to the new `StringUtil` class
- Removed the deprecated `public @Nullable String uriDecode(final @Nullable String source)`
- Update `@since 1.6.0` to `@since 2.0.0` (this was missed in #219)

ppkarwasz commented:
:heavy_plus_sign: 1 for refactoring `String`-related methods into their own methods, but I don't understand why the breaking change is necessary? The only breaking change is the removal of `PackageURL#uriDecode`, I don't think it is worth scaring users with a major release.

ppkarwasz commented on src/main/java/com/github/packageurl/utils/StringUtil.java:
```diff
@@ -0,0 +1,266 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.github.packageurl.utils;
```
Since these classes are not meant to be used by third-party libraries, I would suggest:

- Using `com.github.packageurl.internal` or similar as package name.
- Documenting that the package is internal in `package-info.java`.

Due to #201, this package will **not** be exported through JPMS, since `package-info.java` is not annotated with `@Export`.

ppkarwasz commented on src/main/java/com/github/packageurl/utils/StringUtil.java:
```diff
@@ -0,0 +1,266 @@
+    private static final byte PERCENT_CHAR = '%';
+
+    private StringUtil() {
+        throw new AssertionError("Cannot instantiate StringUtil");
```
_Nit_: Isn't this overkill? The constructor is already `private`.

jeremylong replied:
Yeah - likely overkill.

jeremylong replied:
Every once in a while I'm curious about what ChatGPT or other models say is a best practice. In this case, it recommended this overkill throw. So I put it in when I added the private constructor.

ppkarwasz replied:
From time to time I am curious too. Then I remember that these model are trained on code that was `Ctrl+C/Ctrl+V` from StackOverflow.

ppkarwasz commented on src/main/java/com/github/packageurl/PackageURL.java:
```diff
@@ -248,7 +244,7 @@ public PackageURL(
      * @throws MalformedPackageURLException if parsing fails
      * @throws NullPointerException if {@code type} or {@code name} are {@code null}
-     * @since 1.6.0
+     * @since 2.0.0
      */
     public PackageURL(
```
With the introduction of this constructor, the deprecated constructor that takes `TreeMap` above should also be removed.

jeremylong replied:
I'll remove this in another PR.

jeremylong replied:
This was added to #219.

ppkarwasz commented:
> The only breaking change is the removal of `PackageURL#uriDecode`, I don't think it is worth scaring users with a major release.

Sorry, I just saw the change to `PackageURL#getQualifiers`.

jeremylong commented:
This PR also made minor changes to the `percentDecode` and `percentEncode`. The benchmarks before and after show an improvement in percentEncode. However, I'm guessing there is an issue with the benchmark for percentDecode as the times didn't change and are fast enough I'm guessing only the search for an encoded string was not found.

Benchmark of updated percent decode/encode:

| Benchmark             |            (nonAsciiProb) | Mode | Cnt  |   Score |  | Error | Units |
|-------------------|-----------------------|-------|-----|---------|-|------|------| 
| StringUtilBenchmark.baseline         |            0 | avgt  |  25  |  689.715 | ± | 7.937 |  us/op |
| StringUtilBenchmark.baseline          |         0.1 | avgt  |  25 |  1347.950 | ±|  3.308 |  us/op |
| StringUtilBenchmark.baseline           |        0.5 | avgt  |  25 |  2609.872 | ± | 5.754 |  us/op |
| StringUtilBenchmark.percentDecode |               0 | avgt  |  25 |   191.887 | ± | 0.347 |  us/op |
| StringUtilBenchmark.percentDecode   |           0.1 | avgt  |  25 |   192.141 | ±|  0.208 |  us/op |
| StringUtilBenchmark.percentDecode     |         0.5 | avgt |   25 |   192.031 | ± | 0.190 |  us/op |
| StringUtilBenchmark.percentEncode       |         0 | avgt  |  25 |  1023.150 | ± | 3.235  | us/op |
| StringUtilBenchmark.percentEncode       |       0.1 | avgt  |  25 |  3269.102 | ±|  7.088 |  us/op |
| StringUtilBenchmark.percentEncode        |      0.5 | avgt |   25 |  7179.152 | ± | 7.827 |  us/op |


| Benchmark                 |              (nonAsciiProb)  | Mode  | Cnt     | Score    | | Error |  Units |
|----------------------|------------------------|------|------|------|-|-----|-|
| PercentEncodingBenchmark.baseline          |          0 |  avgt |   25  |  686.520 | ±|   6.698  | us/op |
| PercentEncodingBenchmark.baseline          |         0.1 |  avgt  |  25 |  1344.912 | ± |  3.942 |    us/op |
| PercentEncodingBenchmark.baseline           |        0.5 |  avgt   | 25 |  2614.673 | ± |  3.389  | us/op |
| PercentEncodingBenchmark.percentDecode |               0 |  avgt |   25  |  191.987 | ± |  0.319 |  us/op |
| PercentEncodingBenchmark.percentDecode   |           0.1 |  avgt  |  25  |  192.025 | ± |  0.227 |  us/op |
| PercentEncodingBenchmark.percentDecode     |         0.5 |  avgt  |  25 |   191.950 | ±|   0.293 |  us/op |
| PercentEncodingBenchmark.percentEncode      |          0 |  avgt  |  25 | 1158.468 | ± |  3.644 |  us/op |
| PercentEncodingBenchmark.percentEncode       |       0.1 |  avgt |   25  | 2172.666 | ± |  6.813  | us/op |
| PercentEncodingBenchmark.percentEncode        |      0.5 |  avgt |   25 |  4432.998 | ± | 12.272 |  us/op |

jeremylong commented:
I figured out the problem with the benchmark and I'm re-running it now.

ppkarwasz commented:
> I figured out the problem with the benchmark and I'm re-running it now.

The `setup()` method has a bug (`encodedData = encodeData(encodedData)`), I'll post an improved benchmark soon.

jeremylong commented:
I can update the benchmark as part of this PR.

ppkarwasz commented:
Since `nonAsciiProb == 0` (i.e. there are no characters to encode or to decode) is in practice the most common case, we should probably aggressively optimize for it. The `percentDecode` can be easily optimized by skipping all processing if the are no `%` characters:

```java
if (source.indexOf(PERCENT_CHAR) == -1) {
    return source;
}
```

I am not sure if `percentEncode` can get much better.

jeremylong commented:
After updating to use the new benchmark. You'll notice that there isn't a lot of change in the `percentDecode`. Again, I think I have the solution to this and I'll include it in this PR. I'll be back in 4+ hours (I really need to go buy a better dev machine ;)).

With my changes:
| Benchmark                           | (nonAsciiProb) | Mode | Cnt | Score    | Error    | Units |
|-------------------------------------|----------------|------|-----|----------|----------|-------|
| StringUtilBenchmark.baseline        | 0              | avgt | 25  | 44.430   | ± 0.247  | us/op |
| StringUtilBenchmark.baseline        | 0.1            | avgt | 25  | 44.335   | ± 0.371  | us/op |
| StringUtilBenchmark.baseline        | 0.5            | avgt | 25  | 44.474   | ± 0.256  | us/op |
| StringUtilBenchmark.percentDecode   | 0              | avgt | 25  | 191.667  | ± 0.247  | us/op |
| StringUtilBenchmark.percentDecode   | 0.1            | avgt | 25  | 191.876  | ± 0.110  | us/op |
| StringUtilBenchmark.percentDecode   | 0.5            | avgt | 25  | 191.632  | ± 0.216  | us/op |
| StringUtilBenchmark.percentEncode   | 0              | avgt | 25  | 1012.234 | ± 8.376  | us/op |
| StringUtilBenchmark.percentEncode   | 0.1            | avgt | 25  | 1001.798 | ± 4.785  | us/op |
| StringUtilBenchmark.percentEncode   | 0.5            | avgt | 25  | 993.283  | ± 3.431  | us/op |
| StringUtilBenchmark.toLowerCase     | 0              | avgt | 25  | 97.897   | ± 0.134  | us/op |
| StringUtilBenchmark.toLowerCase     | 0.1            | avgt | 25  | 97.759   | ± 0.232  | us/op |
| StringUtilBenchmark.toLowerCase     | 0.5            | avgt | 25  | 98.030   | ± 0.302  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0              | avgt | 25  | 910.749  | ± 3.538  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0.1            | avgt | 25  | 911.569  | ± 3.500  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0.5            | avgt | 25  | 907.451  | ± 2.940  | us/op |

Legacy version:

| Benchmark                                | (nonAsciiProb) | Mode | Cnt | Score     | Error    | Units |
|------------------------------------------|----------------|------|-----|-----------|----------|-------|
| PercentEncodingBenchmark.baseline        | 0              | avgt | 25  | 44.774    | ± 0.299  | us/op |
| PercentEncodingBenchmark.baseline        | 0.1            | avgt | 25  | 44.258    | ± 0.478  | us/op |
| PercentEncodingBenchmark.baseline        | 0.5            | avgt | 25  | 44.443    | ± 0.321  | us/op |
| PercentEncodingBenchmark.percentDecode   | 0              | avgt | 25  | 191.934   | ± 0.276  | us/op |
| PercentEncodingBenchmark.percentDecode   | 0.1            | avgt | 25  | 192.135   | ± 0.171  | us/op |
| PercentEncodingBenchmark.percentDecode   | 0.5            | avgt | 25  | 191.946   | ± 0.349  | us/op |
| PercentEncodingBenchmark.percentEncode   | 0              | avgt | 25  | 1161.316  | ± 7.246  | us/op |
| PercentEncodingBenchmark.percentEncode   | 0.1            | avgt | 25  | 1147.482  | ± 4.225  | us/op |
| PercentEncodingBenchmark.percentEncode   | 0.5            | avgt | 25  | 1149.934  | ± 5.693  | us/op |
| PercentEncodingBenchmark.toLowerCase     | 0              | avgt | 25  | 97.918    | ± 0.352  | us/op |
| PercentEncodingBenchmark.toLowerCase     | 0.1            | avgt | 25  | 97.995    | ± 0.168  | us/op |
| PercentEncodingBenchmark.toLowerCase     | 0.5            | avgt | 25  | 98.058    | ± 0.355  | us/op |
| PercentEncodingBenchmark.toLowerCaseJre  | 0              | avgt | 25  | 912.258   | ± 1.517  | us/op |
| PercentEncodingBenchmark.toLowerCaseJre  | 0.1            | avgt | 25  | 912.719   | ± 1.942  | us/op |
| PercentEncodingBenchmark.toLowerCaseJre  | 0.5            | avgt | 25  | 914.154   | ± 3.604  | us/op |

ppkarwasz commented:
> I'll be back in 4+ hours (I really need to go buy a better dev machine ;)).

JMH tests have a duration expressed in secodns and do not depend on the machine. :wink:

ppkarwasz commented:
In #224 I expanded on this PR by optimizing the case when no percent encoding is needed.

Profiling has shown that `shouldEncode` is the slowest method.

jeremylong commented:
Okay, I updated the benchmark a bit to get the following before/after:


Updated code:

| Benchmark                           | (nonAsciiProb) | Mode | Cnt | Score    | Error    | Units |
|-------------------------------------|----------------|------|-----|----------|----------|-------|
| StringUtilBenchmark.baseline        | 0              | avgt | 25  | 46.163   | ± 0.651  | us/op |
| StringUtilBenchmark.baseline        | 0.1            | avgt | 25  | 45.743   | ± 0.866  | us/op |
| StringUtilBenchmark.baseline        | 0.5            | avgt | 25  | 45.402   | ± 0.220  | us/op |
| StringUtilBenchmark.percentDecode   | 0              | avgt | 25  | 644.170  | ± 1.077  | us/op |
| StringUtilBenchmark.percentDecode   | 0.1            | avgt | 25  | 643.377  | ± 1.724  | us/op |
| StringUtilBenchmark.percentDecode   | 0.5            | avgt | 25  | 643.688  | ± 1.660  | us/op |
| StringUtilBenchmark.percentEncode   | 0              | avgt | 25  | 1727.554 | ± 5.290  | us/op |
| StringUtilBenchmark.percentEncode   | 0.1            | avgt | 25  | 1725.761 | ± 2.922  | us/op |
| StringUtilBenchmark.percentEncode   | 0.5            | avgt | 25  | 1734.117 | ± 6.954  | us/op |
| StringUtilBenchmark.toLowerCase     | 0              | avgt | 25  | 97.806   | ± 0.236  | us/op |
| StringUtilBenchmark.toLowerCase     | 0.1            | avgt | 25  | 97.664   | ± 0.267  | us/op |
| StringUtilBenchmark.toLowerCase     | 0.5            | avgt | 25  | 97.952   | ± 0.445  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0              | avgt | 25  | 754.331  | ± 7.266  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0.1            | avgt | 25  | 750.950  | ± 6.332  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0.5            | avgt | 25  | 748.199  | ± 6.261  | us/op |

Legacy code:

| Benchmark                                | (nonAsciiProb) | Mode | Cnt | Score    | Error    | Units |
|------------------------------------------|----------------|------|-----|----------|----------|-------|
| PercentEncodingBenchmark.baseline        | 0              | avgt | 25  | 45.611   | ± 0.276  | us/op |
| PercentEncodingBenchmark.baseline        | 0.1            | avgt | 25  | 45.062   | ± 0.352  | us/op |
| PercentEncodingBenchmark.baseline        | 0.5            | avgt | 25  | 45.586   | ± 0.228  | us/op |
| PercentEncodingBenchmark.percentDecode   | 0              | avgt | 25  | 700.983  | ± 2.336  | us/op |
| PercentEncodingBenchmark.percentDecode   | 0.1            | avgt | 25  | 701.126  | ± 4.525  | us/op |
| PercentEncodingBenchmark.percentDecode   | 0.5            | avgt | 25  | 699.940  | ± 2.862  | us/op |
| PercentEncodingBenchmark.percentEncode   | 0              | avgt | 25  | 1596.301 | ± 3.507  | us/op |
| PercentEncodingBenchmark.percentEncode   | 0.1            | avgt | 25  | 1599.507 | ± 3.095  | us/op |
| PercentEncodingBenchmark.percentEncode   | 0.5            | avgt | 25  | 1599.060 | ± 3.003  | us/op |
| PercentEncodingBenchmark.toLowerCase     | 0              | avgt | 25  | 97.740   | ± 0.269  | us/op |
| PercentEncodingBenchmark.toLowerCase     | 0.1            | avgt | 25  | 97.811   | ± 0.207  | us/op |
| PercentEncodingBenchmark.toLowerCase     | 0.5            | avgt | 25  | 98.061   | ± 0.350  | us/op |
| PercentEncodingBenchmark.toLowerCaseJre  | 0              | avgt | 25  | 749.010  | ± 6.381  | us/op |
| PercentEncodingBenchmark.toLowerCaseJre  | 0.1            | avgt | 25  | 750.088  | ± 6.285  | us/op |
| PercentEncodingBenchmark.toLowerCaseJre  | 0.5            | avgt | 25  | 747.762  | ± 6.353  | us/op |

jeremylong commented:
Minor optimization add:

| Benchmark                           | (nonAsciiProb) | Mode | Cnt | Score    | Error    | Units |
|-------------------------------------|----------------|------|-----|----------|----------|-------|
| StringUtilBenchmark.baseline        | 0              | avgt | 25  | 44.112   | ± 0.352  | us/op |
| StringUtilBenchmark.baseline        | 0.1            | avgt | 25  | 865.644  | ± 6.750  | us/op |
| StringUtilBenchmark.baseline        | 0.5            | avgt | 25  | 2544.884 | ± 3.477  | us/op |
| StringUtilBenchmark.percentDecode   | 0              | avgt | 25  | 23.847   | ± 0.075  | us/op |
| StringUtilBenchmark.percentDecode   | 0.1            | avgt | 25  | 961.505  | ± 4.098  | us/op |
| StringUtilBenchmark.percentDecode   | 0.5            | avgt | 25  | 2991.065 | ± 6.832  | us/op |
| StringUtilBenchmark.percentEncode   | 0              | avgt | 25  | 134.289  | ± 3.077  | us/op |
| StringUtilBenchmark.percentEncode   | 0.1            | avgt | 25  | 1245.508 | ± 2.305  | us/op |
| StringUtilBenchmark.percentEncode   | 0.5            | avgt | 25  | 4177.429 | ± 10.390 | us/op |
| StringUtilBenchmark.toLowerCase     | 0              | avgt | 25  | 97.834   | ± 0.113  | us/op |
| StringUtilBenchmark.toLowerCase     | 0.1            | avgt | 25  | 117.929  | ± 0.677  | us/op |
| StringUtilBenchmark.toLowerCase     | 0.5            | avgt | 25  | 111.518  | ± 0.192  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0              | avgt | 25  | 907.524  | ± 3.168  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0.1            | avgt | 25  | 1361.880 | ± 3.625  | us/op |
| StringUtilBenchmark.toLowerCaseJre  | 0.5            | avgt | 25  | 2137.506 | ± 18.792 | us/op |

jeremylong commented:
@ppkarwasz I think this is good to go. I don't think we can get much more optimization out of the encode/decode and moving the string functions to their own class helps clean up the `PackageURL` class.

ppkarwasz commented on src/main/java/com/github/packageurl/internal/package-info.java:
```diff
@@ -0,0 +1,26 @@
+ * SOFTWARE.
+ */
+
+/**
+ * This package contains utility classes used by the PackageURL library.
+ */
+package com.github.packageurl.internal;
```
A `@NullMarked` annotation would be useful here.

jeremylong replied:
I just added the jspecify annotations to both the `package-info` and `StringUtil`.

ppkarwasz reviewed (approved):
LGTM

Around 100 ns per operation on a 256 character long `String` looks good enough to me.

Maybe we could split the `toLowerCase` and `toLowerCaseJre` benchmark method to a separate benchmark class: right now these methods use the test strings for encoding, so there are no favorable test string (e.g. a string with only lowercase characters).
